### ISSUE TITLE
stdio endpoint tests leave a dirty reactor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5730]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)| @ashfall reported|
|-|-|
|Trac ID|trac#5730|
|Type|defect|
|Created|2012-06-28 13:25:58Z|
|Branch|https://github.com/twisted/twisted/tree/dirty-stdio-endpoint-5730|

The problem with many of the tests in `StandardIOEndpointsTestCase` is that they're using a real reactor, but not actually [waiting for the disconnection notification](http://mumak.net/stuff/twisted-disconnect.html).

It would be nice to avoid using a real reactor (or, for that matter, real stdio streams) in most of these tests; if we want an integration test it should probably live in an isolated subprocess.

As per [this](http://twistedmatrix.com/trac/ticket/4697#comment:18) comment on [#4697](https://github.com/twisted/twisted/issues/4697), this needs to be fixed.

<details><summary>Searchable metadata</summary>

```
trac-id__5730 5730
type__defect defect
reporter__ashfall ashfall
priority__normal normal
milestone__ 
branch__branches_dirty_stdio_endpoint_5730 branches/dirty-stdio-endpoint-5730
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1340889958000000 1340889958000000
changetime__1403823331053898 1403823331053898
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_1"></a>|@ashfall commented|
|-|-|

[#5892](https://github.com/twisted/twisted/issues/5892) looks like a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

Previously, this ticket indicated that the issue was with Windows.  However, when running test_endpoints under Emacs (which has a relatively slow pipe consumer on stdout) I have been seeing this somewhat regularly: 

```
[ERROR]
Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
Selectables:
<twisted.internet.process.ProcessWriter object at 0x104216e10>

twisted.internet.test.test_endpoints.StandardIOEndpointsTestCase.test_protocol
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

(In [42774]) Branching to dirty-stdio-endpoint-5730.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban set owner to @glyph|
|-|-|

Does this have a news file?

----------

Is this supposed to work on Python 3, as for PY3 I see that it is set to None?

-----

with then new setup method I find it hard to read tests...
I read the docstring of test_protocolCreation ... and then read
the code... but it looks like protocol is already created and the
test does not directly create the protocol
...same for test_passedReactor...

Maybe this can be merged into a single test

------

I think that the patch looks good and the removal of reactor singleton usage is welcomed.

Thanks

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Replying to [adiroiban](#note_5):
> Does this have a news file?

It does now.  Just a misc though, this is just a test issue.

> Is this supposed to work on Python 3, as for PY3 I see that it is set to None?

There's no change to python 3 support in this branch, as it's just a fix for a test failure.

> with then new setup method I find it hard to read tests...
> I read the docstring of test_protocolCreation ... and then read
> the code... but it looks like protocol is already created and the
> test does not directly create the protocol
> ...same for test_passedReactor...
> 
> Maybe this can be merged into a single test

Out of scope for this change :-).  Also, tests should be as fine-grained as possible, and I don't see that a lot would be gained by merging them together.

> I think that the patch looks good and the removal of reactor singleton usage is welcomed.

Great.

Again, for future reference, please include the next step (please merge, please resubmit) in your reviews.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set status to `closed`|
|-|-|

(In [42819]) Merge dirty-stdio-endpoint-5730

Author: dreid, glyph
Reviewer: adiroiban
Fixes: [#5730](https://github.com/twisted/twisted/issues/5730)

Fix an intermittently failing (dirty reactor) test for StandardIOEndpoint.
